spicy/toolchain/src/compiler/visitors/validator.cc

```diff
@@ -269,6 +270,17 @@ struct PreTransformVisitor : public hilti::visitor::PreOrder<void, PreTransformV
                 error(fmt("%requires expression must be of type bool, but is of type %d ", e->type()), p);
         }
 
+        else if ( i.id().str() == "%context" ) {
+            if ( auto e = i.expression(); ! e )
+                error("%context requires an argument", p);
+            else if ( ! e->isA<hilti::expression::Type_>() )
+                error("%context requires a type", p);
+
+            auto decl = p.findParent<hilti::declaration::Type>();
+            if ( decl && decl->get().linkage() != hilti::declaration::Linkage::Public )
+                error("only public units can have %context", p);
```

Comment:
It would need extra code to create and set those contexts. And conceptually I prefer to restrict them to public units to emphasize that that's what their purpose is.